File: source/go/knapsack.go
package main

import "fmt"

func knapsack(weights, values []int, capacity int) int {
	n := len(weights)
	dp := make([][]int, n+1)
	for i := 0; i <= n; i++ {
		dp[i] = make([]int, capacity+1)
	}

	for i := 1; i <= n; i++ {
		for w := 0; w <= capacity; w++ {
			dp[i][w] = dp[i-1][w]
			if weights[i-1] <= w {
				withItem := dp[i-1][w-weights[i-1]] + values[i-1]
				if withItem > dp[i][w] {
					dp[i][w] = withItem
				}
			}
		}
	}
	return dp[n][capacity]
}

func main() {
	weights := []int{2, 3, 4, 5}
	values := []int{3, 4, 5, 6}
	capacity := 8
	result := knapsack(weights, values, capacity)
	fmt.Printf("max value: %d\n", result)
}

File: source/rust/knapsack.rs
fn knapsack(weights: &[usize], values: &[i32], capacity: usize) -> i32 {
    let n = weights.len();
    let mut dp = vec![vec![0i32; capacity + 1]; n + 1];
    for i in 1..=n {
        for w in 0..=capacity {
            dp[i][w] = dp[i-1][w];
            if weights[i-1] <= w {
                let with_item = dp[i-1][w - weights[i-1]] + values[i-1];
                dp[i][w] = dp[i][w].max(with_item);
            }
        }
    }
    dp[n][capacity]
}

fn main() {
    let weights = [2, 3, 4, 5];
    let values = [3, 4, 5, 6];
    let result = knapsack(&weights, &values, 8);
    println!("max value: {}", result);
}
